### PR TITLE
[Compiler] Optimizer to remove unused instructions

### PR DESCRIPTION
**Pull request by Hecate2**
- I used an individual namespace Neo.Optimizer. This is to prevent conflict between Neo.VM.Instruction and Neo.Compiler.Instruction, and to indicate that Neo.Optimizer optimizes contracts written by all languages, not only C#.
- I passed all tests, but I am not really sure whether my optimizations had been applied...
- I actually used neo-fairy-test to test my codes against [NFTLoan](https://github.com/Hecate2/NFTLoan) (https://github.com/Hecate2/neo-fairy-client/blob/master/test_nftloan.py) & [RelationalDB](https://github.com/Hecate2/RelationalDB/)
- My own tests did not verify rarely-seen Instructions like ENDFINALLY and ASSERTMSG. They may not be handled correctly.

### CODE REVIEW

**Comment by Jim8y** [comment]
Please put your file into a folder.

**Comment by Hecate2** [comment]
> Please put your files into a folder.

WILCO

**Comment by Hecate2** [comment]
There is still bug:
https://github.com/Hecate2/neo-devpack-dotnet/blob/d4523dae603ff25000b0be527360fb482544ec59/src/Neo.Compiler.CSharp/Optimizer/AssemblyOptimizer.cs#L106-L116
I did not remove `methodsToRemove` from debugInfo

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
@superboyiii @shargon @cschuchardt88  please review

File: src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+                    methodsToRemove.Add(method);
+                    continue;
+                }
+                int methodStart = (int)simplifiedInstructionsToAddress[oldAddressToInstruction[oldMethodStart]]!;
+                int methodEnd = (int)simplifiedInstructionsToAddress[oldAddressToInstruction[oldMethodEnd]]!;
+                newMethodStart.Add(methodStart, method["id"]!.AsString());  // TODO: same format of method name as dumpnef
+                newMethodEnd.Add(methodEnd, method["id"]!.AsString());
+                method["range"] = $"{methodStart}-{methodEnd}";
+
+                Regex sequencePointRegex = new Regex(@"(\d+)(\[\d+\]\d+\:\d+\-\d+\:\d+)");
```
Comment (cschuchardt88) [inline src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs:108]: Use compiled regex.

File: src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+                    Instruction oldInstruction = oldAddressToInstruction[startingInstructionAddress];
+                    if (simplifiedInstructionsToAddress.Contains(oldInstruction))
+                    {
+                        startingInstructionAddress = (int)simplifiedInstructionsToAddress[oldInstruction]!;
+                        newSequencePoints.Add(new JString($"{startingInstructionAddress}{sequencePointGroups[2]}"));
+                        previousSequencePoint = startingInstructionAddress;
+                    }
+                    else
+                        newSequencePoints.Add(new JString($"{previousSequencePoint}{sequencePointGroups[2]}"));
+                    Regex documentRegex = new Regex(@"\[(\d+)\](\d+)\:(\d+)\-(\d+)\:(\d+)");
```
Comment (cschuchardt88) [inline src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs:124]: Use Compiled regex.

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]
Please make changes before we can merge. Please keep the coding style the same.

**Comment by Hecate2** [comment]
> Please make changes before we can merge.

OK

**Comment by cschuchardt88** [comment]
Please keep the coding style the same.

File: src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            using var zippedStream = new MemoryStream(zippedBuffer);
+            using var archive = new ZipArchive(zippedStream);
+            var entry = archive.Entries.FirstOrDefault();
+            if (entry != null)
+            {
+                using var unzippedEntryStream = entry.Open();
+                using var ms = new MemoryStream();
+                unzippedEntryStream.CopyTo(ms);
+                var unzippedArray = ms.ToArray();
+                return Encoding.Default.GetString(unzippedArray);
```
[the remark this answers is not in the document]
Comment (Hecate2) [reply src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs:30]: `Zip` and `Unzip` are utility functions and are not used for now. Will delete them.

**Comment by cschuchardt88** [comment]
@Jim8y we dont have the `.editorconfig` on this repo?

**Comment by Jim8y** [comment]
> @Jim8y we dont have the `.editorconfig` on this repo?

You can add it

File: src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs [hunk 4]
```diff
@@ -16,39 +16,31 @@ namespace Neo.Optimizer
...
-        public static string Unzip(byte[] zippedBuffer)
+        static readonly Lazy<IReadOnlyDictionary<uint, string>> sysCallNames = new(
+            () => ApplicationEngine.Services.ToImmutableDictionary(kvp => kvp.Value.Hash, kvp => kvp.Value.Name));
+
+        public static string Unzip(string path)
         {
-            using var zippedStream = new MemoryStream(zippedBuffer);
-            using var archive = new ZipArchive(zippedStream);
+            using FileStream zippedBuffer = File.OpenRead(path);
+            using var archive = new ZipArchive(zippedBuffer);
```
Comment (cschuchardt88) [inline src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs:25]: You may want to do `ZipArchive(zippedBuffer, Encoding.UTF8, false);` to ensure that encoding for the `GetString` method below.

**Review by cschuchardt88 — COMMENTED** [review]
Looks good.

File: src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs [hunk 5]
```diff
@@ -16,39 +16,31 @@ namespace Neo.Optimizer
...
             var entry = archive.Entries.FirstOrDefault();
             if (entry != null)
             {
                 using var unzippedEntryStream = entry.Open();
                 using var ms = new MemoryStream();
                 unzippedEntryStream.CopyTo(ms);
                 var unzippedArray = ms.ToArray();
-                return Encoding.Default.GetString(unzippedArray);
+                return Encoding.UTF8.GetString(unzippedArray);
```
[the remark this answers is not in the document]
Comment (Hecate2) [reply src/Neo.Compiler.CSharp/Optimizer/DumpNef.cs:33]: Zip and Unzip methods are used for debug info (.nefdbgnfo) files. The debug info is a zip archive, containing a single .debug.json file inside.

**Review by cschuchardt88 — APPROVED** [review]

**Comment by Jim8y** [comment]
@shargon

File: tests/Neo.SmartContract.TestEngine/TestEngine.cs [hunk 6]
```diff
@@ -52,15 +52,21 @@ public TestEngine(TriggerType trigger = TriggerType.Application, IVerifiable? ve
...
+            return AddEntryScript(true, true, files);
         }
 
-        public CompilationContext AddEntryScript(bool debug = true, params string[] files)
+        public CompilationContext AddNoOptimizeEntryScript(params string[] files)
+        {
+            return AddEntryScript(false, true, files);
+        }
+
+        public CompilationContext AddEntryScript(bool optimize = true, bool debug = true, params string[] files)
```
Comment (shargon) [inline tests/Neo.SmartContract.TestEngine/TestEngine.cs:63]: Maybe false by default...

**Review by shargon — COMMENTED** [review]
> I passed all tests, but I am not really sure whether my optimizations had been applied...

Could you create a unit test for each optimization?

> My own tests did not verify rarely-seen Instructions like ENDFINALLY and ASSERTMSG. They may not be handled correctly.

This is scary, could you add a unit test to ensure that this blocks are well optimized?

**Comment by shargon** [comment]
I will revert it, sorry

**Comment by shargon** [comment]
@Hecate2 could you create a new pr for the unit tests?

**Comment by Hecate2** [comment]
> @Hecate2 could you create a new pr for the unit tests?

OK. Yet I am still wondering how to test huge contracts in C# running quickly in github actions.

**Comment by shargon** [comment]
I have in mind a way, not only for c#, for all nef contracts to be tested easily

**Comment by shargon** [comment]
@Hecate2 I can confirm that it didn't work with `TRY` blocks could you take a look or avoid it?

**Comment by cschuchardt88** [comment]
> > @Hecate2 could you create a new pr for the unit tests?
> 
> OK. Yet I am still wondering how to test huge contracts in C# running quickly in github actions.

`neo-express` has a tool library for `XUnit` https://github.com/neo-project/neo-express/tree/master/src/test-harness. We use it in the `neo-express` github actions I believe. However it uses `neo-express` node to execute contract.

Here's how to use. https://github.com/cschuchardt88/neo-examples-csharp/tree/master/tests/Contract.Tests

maybe you can code up something.

**Comment by Hecate2** [comment]
> @Hecate2 I can confirm that it didn't work with `TRY` blocks could you take a look or avoid it?

I found the problem. Working on the PR

**Pull request merged**